applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+	struct net_buf *buf;
+
+	__ASSERT_NO_MSG(p_chmap != NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+				sizeof(*cp));
+	if (!buf) {
+		return -1;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	if (!cp) {
+		return -1;
+	}
+
+	memcpy(cp->ch_map, p_chmap, sizeof(cp->ch_map));
+	__ASSERT_NO_MSG(sizeof(cp->ch_map) == 5);
+
+	return bt_hci_cmd_send_sync(
+			BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+			buf,
+			NULL);
+}
+
+static bool on_vs_evt(struct net_buf_simple *buf)
+{
+	u8_t *p_subevent_code;
+	hci_vs_evt_qos_conn_event_report_t *p_evt;
+
+	p_subevent_code = net_buf_simple_pull_mem(
+						buf,
+						sizeof(*p_subevent_code));
+
+	switch (*p_subevent_code) {
+	case HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT:
+	{
+		if (atomic_get(&m_processing) == 1) {
+			/* Cheaper to skip this update */
+			/* instead of using locks */
+			return true;
+		}
+
+		p_evt = (void *)buf->data;
+
+		chmap_filter_crc_update(
+			mp_chmap_inst,
+			p_evt->channel_index,
+			p_evt->crc_ok_count,
+			p_evt->crc_error_count);
+		return true;
+	}
+	default:
+		return false;
+	}
+}
+
+static void enable_qos_reporting(void)
+{
+	int err;
+	struct net_buf *buf;
+
+	err = bt_hci_register_vnd_evt_cb(on_vs_evt);
+	if (err) {
+		LOG_ERR("Failed to register HCI VS callback");
+		return;
+	}
+
+	hci_vs_cmd_qos_conn_event_report_enable_t *p_cmd_enable;
+
+	buf = bt_hci_cmd_create(HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+							sizeof(*p_cmd_enable));
+	if (!buf) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+
+	p_cmd_enable = net_buf_add(buf, sizeof(*p_cmd_enable));
+	p_cmd_enable->enable = 1;
+
+	err = bt_hci_cmd_send(
+		HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+		buf);
+	if (err) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	switch (OPT_FIELD_GET(event->id)) {
+	case QOS_OPT_BLACKLIST:
+		update_blacklist(
+			(struct params_blacklist *)event->dyndata.data);
+		break;
+	case QOS_OPT_CHMAP:
+		/* Avoid updating channel map directly to */
+		/* reduce complexity in interaction with */
+		/* chmap filter lib */
+		LOG_DBG("Not supported");
+		break;
+	case QOS_OPT_PARAM_BLE:
+		__ASSERT_NO_MSG(
+			event->dyndata.size == sizeof(struct params_ble));
+		update_parameters(
+			(struct params_ble *)event->dyndata.data,
+			NULL);
+		break;
+	case QOS_OPT_PARAM_WIFI:
+		__ASSERT_NO_MSG(
+			event->dyndata.size == sizeof(struct params_wifi));
+		update_parameters(
+			NULL,
+			(struct params_wifi *)event->dyndata.data);
+		break;
+	default:
+		LOG_WRN("Unknown opt");
+		return;
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event;
+
+		event = cast_module_state_event(eh);
+
+		if (check_state(
+				event,
+				MODULE_ID(ble_state),
+				MODULE_STATE_READY)) {
+			static bool initialized;
+
+			__ASSERT_NO_MSG(!initialized);
```

Comment:
> You're suggesting to use `MODULE_ID(main) ` instead of `MODULE_ID(ble_state)`?

Only for kernel and mutex initialization. (as well as other runtime, one-time initialized things). Also assert for initialized should be moved there.

Everything that is dependent on BLE being ready should be initialized as is.

> I mirrored the behavior from ble_scan here

@MarekPieta It would be good to review this part. ble module is initialized only once now but it may change if anybody will add ble_disable or something ;)